File: Commons/Graph/src/main/java/com/greentree/commons/graph/algorithm/walk/GraphWalker.java
package com.greentree.commons.graph.algorithm.walk;

import com.greentree.commons.graph.Graph;

public interface GraphWalker<V> {
	
	Graph<? extends V> graph();
	
	
	default void visit(GraphVisitor<? super V> visitor) {
		//		final var used_all = new HashSet<V>();
		//		final var bv = new ProxyGraphVisitor<V>(visitor) {
		//			
		//			@Override
		//			public boolean startVisit(V parent, V v) {
		//				used_all.add(v);
		//				return super.startVisit(parent, v);
		//			}
		//			
		//			@Override
		//			public boolean startVisit(V v) {
		//				used_all.add(v);
		//				return super.startVisit(v);
		//			}
		//			
		//		};
		//		for(var v : graph()) {
		//			if(!used_all.contains(v))
		//				visit(v, bv);
		//		}
		for(var v : graph())
			visit(v, visitor);
	}
	
	void visit(V start, GraphVisitor<? super V> visitor);
	
}

File: Commons/Graph/src/main/java/com/greentree/commons/graph/algorithm/walk/GraphWalker.kt
package com.greentree.commons.graph.algorithm.walk

import com.greentree.commons.graph.FiniteGraph
import com.greentree.commons.graph.algorithm.walk.path.PathToVertexVisitor
import com.greentree.commons.graph.algorithm.walk.path.PathVisitor
import com.greentree.commons.graph.algorithm.walk.vertex.VertexVisitor

interface GraphWalker<out V : Any> {

	fun visit(visitor: VertexVisitor<V>) {
		for(v in finiteGraph!!)
			visit(v, visitor)
	}

	fun visit(visitor: PathVisitor<V>) {
		for(v in finiteGraph!!)
			visit(v, visitor)
	}

	val finiteGraph: FiniteGraph<V>?
		get() = null

	fun visit(start: @UnsafeVariance V, visitor: VertexVisitor<V>)

	fun visit(start: @UnsafeVariance V, visitor: PathVisitor<V>) {
		visit(start, PathToVertexVisitor(visitor))
	}
}
